Lib/xml/dom/minidom.py

```diff
@@ -300,12 +300,28 @@ def _in_document(node):
         node = node.parentNode
     return False
 
-def _write_data(writer, data):
+def _write_data(writer, text, attr):
     "Writes datachars to writer."
-    if data:
-        data = data.replace("&", "&amp;").replace("<", "&lt;"). \
-                    replace("\"", "&quot;").replace(">", "&gt;")
-        writer.write(data)
+    if not text:
+        return
+    # See the comments in ElementTree.py for behavior and
+    # implementation details.
+    if "&" in text:
+        text = text.replace("&", "&amp;")
+    if "<" in text:
+        text = text.replace("<", "&lt;")
+    if ">" in text:
+        text = text.replace(">", "&gt;")
+    if attr:
+        if '"' in text:
+            text = text.replace('"', "&quot;")
+        if "\r" in text:
+            text = text.replace("\r", "&#13;")
+        if "\n" in text:
+            text = text.replace("\n", "&#10;")
+        if "\t" in text:
+            text = text.replace("\t", "&#09;")
```

Comment:
I wonder, why `&#09;` and not simply `&#9;`? Is there a reason?

Reply:
No technical reason. Probably based on Python's hex character spelling or due to consistency with the other two-digit codes above. The XML character spec does not need (or mention) leading zeros.

I'm happy to keep the leading zero. If you need compact data, use compression. That's way more effective than stripping some zeros from rare tab characters.

Reply:
`xml.sax.saxutils.quoteattr()` uses `&#9;`.

Reply:
Which makes me wonder why we need a new implementation here, rather than importing the existing one.

Reply:
I now looked up the implementation in `saxutils.py` – it looks fairly slow. `minidom` will probably not become high-performance by any accident, but it doesn't feel good to  slow it down even more. It's probably worth a new implementation..

Reply:
It is sad, but there is a copy of escaping function in almost every module which outputs XML or HTML. On one hand, it is a trivial function, and we want to avoid unneeded dependencies. On other hand, efficient and complete implementation is not so trivial. But `xml.sax.saxutils` version is too generalized and far from been efficient.

It was worse in the past. Now many code just use `html.escape()`.